File: hooks/scripts/cmd/task_completed.go
// TaskCompleted: Log task completion and inject progress context.
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"solon-hooks/internal/config"
	"solon-hooks/internal/hookio"
)

var taskCompletedCmd = &cobra.Command{
	Use:   "task-completed",
	Short: "Handle TaskCompleted hook",
	RunE:  runTaskCompleted,
}

// taskFile is the minimal shape of a ~/.claude/tasks/{team}/*.json file.
type taskFile struct {
	ID     string `json:"id"`
	Status string `json:"status"`
}

// taskCounts holds task status tallies for a team.
type taskCounts struct {
	Pending    int
	InProgress int
	Completed  int
	Total      int
}

// countTeamTasks reads task JSON files and tallies status counts.
func countTeamTasks(teamName string) *taskCounts {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil
	}
	taskDir := filepath.Join(home, ".claude", "tasks", teamName)
	entries, err := os.ReadDir(taskDir)
	if err != nil {
		return nil
	}
	counts := &taskCounts{}
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		data, err := os.ReadFile(filepath.Join(taskDir, e.Name()))
		if err != nil {
			continue
		}
		var t taskFile
		if err := json.Unmarshal(data, &t); err != nil || t.Status == "" {
			continue
		}
		switch t.Status {
		case "pending":
			counts.Pending++
		case "in_progress":
			counts.InProgress++
		case "completed":
			counts.Completed++
		}
	}
	counts.Total = counts.Pending + counts.InProgress + counts.Completed
	return counts
}

// logTaskCompletion appends a completion line to the team reports file.
func logTaskCompletion(teamName, taskID, taskSubject, teammateName string) {
	reportsPath := os.Getenv("SL_REPORTS_PATH")
	if reportsPath == "" {
		return
	}
	logFile := filepath.Join(reportsPath, fmt.Sprintf("team-%s-completions.md", teamName))
	if err := os.MkdirAll(filepath.Dir(logFile), 0755); err != nil {
		return
	}
	timestamp := time.Now().Format("2006-01-02 15:04:05")
	line := fmt.Sprintf("- [%s] Task #%s \"%s\" completed by %s\n", timestamp, taskID, taskSubject, teammateName)
	f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	_, _ = f.WriteString(line)
}

func runTaskCompleted(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("task-completed-handler") {
		os.Exit(0)
	}

	defer func() {
		if r := recover(); r != nil {
			os.Exit(0)
		}
	}()

	var payload hookio.TaskCompletedInput
	if err := hookio.ReadInput(&payload); err != nil {
		os.Exit(0)
	}

	if payload.TeamName == "" {
		os.Exit(0)
	}

	logTaskCompletion(payload.TeamName, payload.TaskID, payload.TaskSubject, payload.TeammateName)

	counts := countTeamTasks(payload.TeamName)
	lines := []string{
		"## Task Completed",
		fmt.Sprintf("Task #%s \"%s\" completed by %s.", payload.TaskID, payload.TaskSubject, payload.TeammateName),
	}

	if counts != nil {
		remaining := counts.Pending + counts.InProgress
		lines = append(lines, fmt.Sprintf(
			"Progress: %d/%d done. %d pending, %d in progress.",
			counts.Completed, counts.Total, counts.Pending, counts.InProgress,
		))
		if remaining == 0 {
			lines = append(lines, "")
			lines = append(lines, "**All tasks completed.** Consider shutting down teammates and synthesizing results.")
		}
	}

	hookio.WriteOutput(map[string]interface{}{
		"hookSpecificOutput": map[string]interface{}{
			"hookEventName":     "TaskCompleted",
			"additionalContext": strings.Join(lines, "\n"),
		},
	})
	return nil
}

File: crates/solon-core/src/hooks/task_completed.rs
use crate::hooks::{get_str, is_hook_enabled};
/// TaskCompleted hook: log task completion and inject progress context.
use anyhow::Result;

pub fn run() -> Result<()> {
    if !is_hook_enabled("task-completed-handler") {
        std::process::exit(0);
    }

    let input = match crate::hooks::read_hook_input() {
        Ok(v) => v,
        Err(_) => std::process::exit(0),
    };

    let team_name = get_str(&input, "team_name").to_string();
    if team_name.is_empty() {
        std::process::exit(0);
    }

    let task_id = get_str(&input, "task_id").to_string();
    let task_subject = get_str(&input, "task_subject").to_string();
    let teammate_name = get_str(&input, "teammate_name").to_string();

    log_task_completion(&team_name, &task_id, &task_subject, &teammate_name);

    let counts = count_team_tasks(&team_name);
    let mut lines = vec![
        "## Task Completed".to_string(),
        format!(
            "Task #{} \"{}\" completed by {}.",
            task_id, task_subject, teammate_name
        ),
    ];

    if let Some(c) = counts {
        let remaining = c.pending + c.in_progress;
        lines.push(format!(
            "Progress: {}/{} done. {} pending, {} in progress.",
            c.completed, c.total, c.pending, c.in_progress
        ));
        if remaining == 0 {
            lines.push(String::new());
            lines.push("**All tasks completed.** Consider shutting down teammates and synthesizing results.".to_string());
        }
    }

    crate::hooks::write_output(&serde_json::json!({
        "hookSpecificOutput": {
            "hookEventName": "TaskCompleted",
            "additionalContext": lines.join("\n"),
        }
    }));
    Ok(())
}

struct TaskCounts {
    pending: usize,
    in_progress: usize,
    completed: usize,
    total: usize,
}

/// Read ~/.claude/tasks/{team}/*.json and tally status counts.
fn count_team_tasks(team_name: &str) -> Option<TaskCounts> {
    let home = dirs::home_dir()?;
    let task_dir = home.join(".claude").join("tasks").join(team_name);
    let entries = std::fs::read_dir(&task_dir).ok()?;

    let mut counts = TaskCounts {
        pending: 0,
        in_progress: 0,
        completed: 0,
        total: 0,
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        let data = match std::fs::read_to_string(&path) {
            Ok(d) => d,
            Err(_) => continue,
        };
        let v: serde_json::Value = match serde_json::from_str(&data) {
            Ok(v) => v,
            Err(_) => continue,
        };
        match v.get("status").and_then(|s| s.as_str()).unwrap_or("") {
            "pending" => counts.pending += 1,
            "in_progress" => counts.in_progress += 1,
            "completed" => counts.completed += 1,
            _ => continue,
        }
    }
    counts.total = counts.pending + counts.in_progress + counts.completed;
    Some(counts)
}

/// Append a completion line to the team reports log file.
fn log_task_completion(team_name: &str, task_id: &str, task_subject: &str, teammate_name: &str) {
    let reports_path = match std::env::var("SL_REPORTS_PATH") {
        Ok(p) if !p.is_empty() => p,
        _ => return,
    };
    let log_file =
        std::path::PathBuf::from(&reports_path).join(format!("team-{}-completions.md", team_name));
    if let Some(parent) = log_file.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    let timestamp = chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
    let line = format!(
        "- [{}] Task #{} \"{}\" completed by {}\n",
        timestamp, task_id, task_subject, teammate_name
    );
    use std::io::Write;
    if let Ok(mut f) = std::fs::OpenOptions::new()
        .append(true)
        .create(true)
        .open(&log_file)
    {
        let _ = f.write_all(line.as_bytes());
    }
}
